src/borg/key.py

```diff
@@ -154,13 +156,82 @@ class AESKeyBase(KeyBase):
+            self.save_repo_end_of_nonce_reservation = lambda next_unreserved, start_nonce: None
+            return None
+        if repo_free_nonce is not None:
+            repo_free_nonce = bytes_to_long(repo_free_nonce)
+        return repo_free_nonce
+
+    def commit_repo_nonce_reservation(self, next_unreserved, start_nonce):
+        if start_nonce is None:
+            self.repository.commit_nonce_reservation(long_to_bytes(next_unreserved), None)
+        else:
+            self.repository.commit_nonce_reservation(long_to_bytes(next_unreserved), long_to_bytes(start_nonce))
+
+    def ensure_reservation(self, nonce_space_needed):
+        if not self.end_of_nonce_reservation:
+            repo_free_nonce = self.get_repo_free_nonce()
+            local_free_nonce = self.get_local_free_nonce()
+            next_nonce = max(x for x in (repo_free_nonce, local_free_nonce, self.manifest_nonce) if x is not None)
+            reservation_end = next_nonce + nonce_space_needed + NONCE_SPACE_RESERVATION
+            self.commit_repo_nonce_reservation(reservation_end, repo_free_nonce)
+            self.commit_local_nonce_reservation(reservation_end, local_free_nonce)
+            self.end_of_nonce_reservation = reservation_end
+            if self.enc_cipher.iv != PREFIX + long_to_bytes(next_nonce):
+                self.enc_cipher.reset(self.enc_key, PREFIX + long_to_bytes(next_nonce))
+        else:
+            next_nonce = bytes_to_long(self.enc_cipher.iv[8:])
+            assert(next_nonce <= self.end_of_nonce_reservation)
+            if next_nonce + nonce_space_needed >= self.end_of_nonce_reservation:
+                # try to expand reservation
+                repo_free_nonce = self.get_repo_free_nonce()
+                local_free_nonce = self.get_local_free_nonce()
+                free_nonce_space = max(x for x in (repo_free_nonce, local_free_nonce, self.manifest_nonce, self.end_of_nonce_reservation) if x is not None)
+                if free_nonce_space == self.end_of_nonce_reservation:
+                    # nothing else got an reservation, just extend
+                    reservation_end = self.end_of_nonce_reservation + nonce_space_needed + NONCE_SPACE_RESERVATION
+                    self.commit_repo_nonce_reservation(reservation_end, repo_free_nonce)
+                    self.commit_local_nonce_reservation(reservation_end, local_free_nonce)
+                    self.end_of_nonce_reservation = reservation_end
+                else:
+                    # need an entirly new reservation
+                    reservation_end = free_nonce_space + nonce_space_needed + NONCE_SPACE_RESERVATION
+                    self.commit_repo_nonce_reservation(reservation_end, repo_free_nonce)
+                    self.commit_local_nonce_reservation(reservation_end, local_free_nonce)
+                    self.end_of_nonce_reservation = reservation_end
+                    self.enc_cipher.reset(self.enc_key, PREFIX + long_to_bytes(next_nonce))
+
```

Comment:
somehow thinking whether this whole nonce mgmt should go into a NonceManager class.